### PR TITLE
shared: vectorize parse_field()

### PR DESCRIPTION
**Pull request by vcaputo**
By vectorizing parse_field() the chain of parse_field() calls in
output_short() can be replaced with a single call receiving a description
of the desired fields and their targets.

While at it, eliminate the repeated strlen() calls performed on constant
field names by making parse_field() receive the field length, and storing
it in the ParseFieldVec at compile time.

Also sort the output_short() fields so the short ones are tried first, for
a minor efficiency gain.

In addition to making the code less repetitive, gcc in my tests now inlines
the parse_fieldv() call in output_short().

### CODE REVIEW

File: src/shared/logs-show.c [hunk 1]
```diff
@@ -500,7 +489,7 @@ static int output_verbose(
         else {
                 _cleanup_free_ char *value = NULL;
 
-                r = parse_field(data, length, "_SOURCE_REALTIME_TIMESTAMP=", &value, NULL);
+                r = parse_field_single(data, length, "_SOURCE_REALTIME_TIMESTAMP=", 27, &value, NULL);
```
Comment (keszybz) [inline src/shared/logs-show.c]: Please use `strlen( "_SOURCE_REALTIME_TIMESTAMP=")` instead. It is replaced by the actual length even at -O0.

File: src/shared/logs-show.c [hunk 2]
```diff
@@ -337,6 +363,17 @@ static int output_short(
         size_t hostname_len = 0, identifier_len = 0, comm_len = 0, pid_len = 0, fake_pid_len = 0, message_len = 0, realtime_len = 0, monotonic_len = 0, priority_len = 0;
         int p = LOG_INFO;
         bool ellipsized = false;
+        ParseFieldVec fields[] = {
```
Comment (keszybz) [inline src/shared/logs-show.c]: `const`

**Comment by vcaputo** [comment]
@keszybz addressed your nits, thanks for the review.

File: src/shared/logs-show.c [hunk 3]
```diff
@@ -79,37 +79,63 @@ static int print_catalog(FILE *f, sd_journal *j) {
         return 0;
 }
 
-static int parse_field(const void *data, size_t length, const char *field, char **target, size_t *target_size) {
-        size_t fl, nl;
+static int parse_field_single(const void *data, size_t length, const char *field, size_t field_len, char **target, size_t *target_len) {
```
Comment (dnicolodi) [inline src/shared/logs-show.c]: Bikeshedding: `parse_one_field()` sounds much better to me.

File: src/shared/logs-show.c [hunk 4]
```diff
@@ -79,37 +79,63 @@ static int print_catalog(FILE *f, sd_journal *j) {
...
 
-        if (target_size)
-                *target_size = nl;
+        if (target_len)
+                *target_len = nl;
 
         return 1;
 }
 
+typedef struct ParseFieldVec {
```
Comment (dnicolodi) [inline src/shared/logs-show.c:112]: Why the `Vec` suffix (standing for "vector" I guess)? This represents a single field. This seems to be used only locally, so I would go for a simple `Field`.

File: src/shared/logs-show.c [hunk 5]
```diff
@@ -79,37 +79,63 @@ static int print_catalog(FILE *f, sd_journal *j) {
...
+        const char *field;
+        size_t field_len;
+        char **target;
+        size_t *target_len;
+} ParseFieldVec;
+
+#define PARSE_FIELD_VEC_ENTRY(_field, _target, _target_len) \
+        { .field = _field, .field_len = strlen(_field), .target = _target, .target_len = _target_len }
+
+static int parse_field(const void *data, size_t length, const ParseFieldVec *fields, unsigned n_fields) {
```
Comment (dnicolodi) [inline src/shared/logs-show.c]: Maybe `parse_fields()` if this now takes an array of fields to parse.

**Comment by vcaputo** [comment]
@dnicolodi Now the vectorized version is simply named `parse_fieldv()` and original name unchanged.  Inspiration for the naming is loosely coming from `writev()` / `struct iovec`.

**Comment by dnicolodi** [comment]
@vcaputo Naming these after something else poorly named does not make them more right :-) But being an internal interface I don't think the bikesheeding is worth the time.

**Comment by vcaputo** [comment]
@dnicolodi Precedent matters, an experienced C programmer looks at `parse_field()` and `parse_fieldv()` and very likely thinks vector.  But hey, for this to really be bikeshedding it requires more people to put in their $.02.

And yes, it's internal, which is why I didn't make much effort sweating the naming in the first place.

**Pull request merged**